### PR TITLE
OpenSDS Profile properties design and implementation

### PR DESCRIPTION
**Pull request by leonwanghui**
**What this PR does / why we need it**:
This is a big patch for implementing [#10](https://github.com/opensds/design-specs/pull/10) to enhance the profile properties design of OpenSDS.

Besides, this patch also removes some redundant code of fake client configuration.

**Which issue this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*: fixes #
design-specs [#10](https://github.com/opensds/design-specs/pull/10)

## Scheduler guidance
The big change of this patch is to update the rules of mapping profiles with available storage pools, and generally the idea is to generate a `filterRequest` stuff according to the definition of `ProfileSpec`. Based on the design of general filter mechanism (thanks to @BaiHuoYu ), I inserted some fields defined in `xxxPropertiesSpec` into `filterRequest`, and then compare all k-v pairs with `Extras` field defined in `StoragePoolSpec`. Although the process is simple, but for nested capability filed the format must be kept as `extras.xxxLoS.{the_field_name}` so that the filter mechanism can work well.

**Special notes for your reviewer**:

**Release note**:

```release-note
Enhances the profile properties design to map profiles with available storage pools.
```

### CODE REVIEW

**Comment by leonwanghui** [comment]
Hi @xing-yang @wisererik @BaiHuoYu , this feature has been implemented completely, please take a review, thx!

File: COMMUNITY.md [hunk 1]
```diff
@@ -1,6 +1,6 @@
 # OpenSDS Controller Work Group
 
-Covers discussion of OpenSDS Controller architecture design and featrue development.
+Covers discussion of OpenSDS Controller architecture design and feature development.
```
Comment (xing-yang) [inline COMMUNITY.md]: Why do you include the documentation change as part of this profiles design PR? Is this a rebase issue?  If you made these changes here, can you separate them into a different PR?

File: openapi-spec/swagger.yaml [hunk 2]
```diff
@@ -521,7 +522,7 @@ paths:
         - name: body
           in: body
           schema:
-            $ref: '#/definitions/ExtendVolumeSpec'
+            $ref: '#/definitions/ExtendVolume'
```
Comment (xing-yang) [inline openapi-spec/swagger.yaml]: API says "resize", but description and schema says "extend".  Not consistent. I'd prefer API is changed to "extend".  Just reminded me of this.  You don't need to change anything as part of this PR.

File: openapi-spec/swagger.yaml [hunk 3]
```diff
@@ -813,7 +814,7 @@ paths:
           description: Forbidden
         '500':
           $ref: '#/responses/HTTPStatus500'
-  '/v1beta/{projectId}/block/volumeGroup/{volumeGroupId}':
+  '/v1beta/{projectId}/block/volumeGroups/{volumeGroupId}':
```
Comment (xing-yang) [inline openapi-spec/swagger.yaml]: This is better changed as a separate PR.  
It's ok to leave it here now.

File: pkg/api/router.go [hunk 4]
```diff
@@ -96,8 +96,8 @@ func Run(host string) {
 				beego.NSRouter("/replications/:replicationId/disable", NewReplicationPortal(), "post:DisableReplication"),
 				beego.NSRouter("/replications/:replicationId/failover", NewReplicationPortal(), "post:FailoverReplication"),
 				// Volume group contains a list of volumes that are used in the same application.
-				beego.NSRouter("/volumeGroup", &VolumeGroupPortal{}, "post:CreateVolumeGroup;get:ListVolumeGroups"),
-				beego.NSRouter("/volumeGroup/:groupId", &VolumeGroupPortal{}, "put:UpdateVolumeGroup;get:GetVolumeGroup;delete:DeleteVolumeGroup"),
+				beego.NSRouter("/volumeGroups", &VolumeGroupPortal{}, "post:CreateVolumeGroup;get:ListVolumeGroups"),
+				beego.NSRouter("/volumeGroups/:groupId", &VolumeGroupPortal{}, "put:UpdateVolumeGroup;get:GetVolumeGroup;delete:DeleteVolumeGroup"),
```
Comment (xing-yang) [inline pkg/api/router.go]: Here too.  It's better to fix this separately.

File: pkg/controller/controller.go [hunk 5]
```diff
@@ -64,11 +63,26 @@ type Controller struct {
 
 func (c *Controller) CreateVolume(ctx *c.Context, in *model.VolumeSpec, errchanVolume chan error) {
 	var err error
-	var profile *model.ProfileSpec
+	var prf *model.ProfileSpec
```
Comment (xing-yang) [inline pkg/controller/controller.go]: I'd prefer to spell it out as "profile", unless if there's any conflict.

File: COMMUNITY.md (hunk 1 above)
Comment (leonwanghui) [reply COMMUNITY.md]: I just wanted to fix this typo error, it's ok to separate it into another PR.

File: openapi-spec/swagger.yaml (hunk 2 above)
Comment (leonwanghui) [reply openapi-spec/swagger.yaml]: Got it, I will assign this issue to @BaiHuoYu

File: openapi-spec/swagger.yaml (hunk 3 above)
Comment (leonwanghui) [reply openapi-spec/swagger.yaml]: @lasshpanzer Please submit another PR to fix it.

File: pkg/api/router.go (hunk 4 above)
Comment (leonwanghui) [reply pkg/api/router.go]: @lasshpanzer Same at here.

File: pkg/controller/controller.go (hunk 5 above)
Comment (leonwanghui) [reply pkg/controller/controller.go]: The reason I modified here is that we usually use `profile` to represent the instance itself, but for the point to the instance, we'd better to use other words.

File: CONTRIBUTING.md [hunk 6]
```diff
@@ -8,35 +8,43 @@
 
 ## How to contribute
 
-opensds is Apache 2.0 licensed and accepts contributions via GitHub pull requests. This document outlines some of the conventions on commit message formatting, contact points for developers and other resources to make getting your contribution into opensds easier.
+opensds is Apache 2.0 licensed and accepts contributions via GitHub pull
+requests. This document outlines some of the conventions on commit message
+formatting, contact points for developers and other resources to make getting
+your contribution into opensds easier.
```
Comment (xing-yang) [inline CONTRIBUTING.md]: @leonwanghui You said you have removed the doc changes, but there are still here?

**Comment by xing-yang** [comment]
@leonwanghui Please remove all the documentation changes from this PR.  They are not related to the profiles properties.  Thanks.
